01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+# Summary
+
+Introduces a new resource interface with the following goals:
+
+* Introduce versioning to the resource interface, so that we can maintain
+  backwards-compatibility.
+
+* Support for spaces (concourse/concourse#1707).
+
+* Introduce a more airtight "versioned artifacts" interface, tightening up
+  loopholes in today's resource API to ensure that resources are pointing at an
+  external source of truth and cannot be partially implemented or hacky.
+
+* Introduce a "notifier" interface. This is to replace a whole class of resource
+  types that will not be able to fit in to an "artifact" interface.
+
+* Extend the versioned artifacts interface to support deletion of versions,
+  either by an explicit `delete` call or by somehow noticing that versions have
+  disappeared.
+
+
+# Proposal
+
+At this early stage of the RFC, it's easiest for me to just use Elm syntax and
+pretend this is all in a type system.
+
+In reality, each of these functions would be scripts with JSON requests passed
+to them on stdin. I'm going to use the `Bits` type just to represent which calls
+have state on disk to either access (like `put`) or return (like `get`). This
+would normally just be the working directory of the script.
+
+## General Types
+
+```elm
+-- arbitrary configuration
+type alias Config = Dict String Json.Value
+
+-- identifier for space, i.e. {"branch":"foo"}
+type alias Space = Dict String String
+
+-- identifier for version, i.e. {"version":"1.2"}
+type alias Version = Dict String String
+
+-- arbitrary ordered metadata (we may make this fancier in the future)
+type alias Metadata = List (String, String)
+
+-- data on disk
+type alias Bits = ()
+```
+
+## Versioned Artifacts interface
+
+```elm
+spaces  : Config -> Set Space
+check   : Config -> Space -> Maybe Version -> List (Version, Metadata)
```

Comment:
```
Spaces are checked independently and have their own version history, 
so this should work fine. You would represent each state as its own space, 
if they can be independently updated.
```
From the elm code and your comment above I understand this as saying that `check` will be passed a specific space to check for new versions? So, for the `git-resource` Concourse would:

1. Ask it for a list of `spaces`, which would then return all the branches in the repository.
2. For each space returned by `spaces`, it would call `check` for that specific space.

This means that the `git-resource` could never benefit from using the Github V4 API (GraphQL), which could return the last commit for all branches in a single API call - since it would essentially be forced to first list all branches, and then iterate over them to find the last commit.

If I understood it correctly, it does not sound ideal?

Reply:
@itsdalmo Interesting point. (Splitting hairs, but I wouldn't expect `git-resource` to hit the GitHub API, since it's generic `git` and not `github`, but your point still stands since there could be some GitHub-specific resource type that does.)

I went with one-check-per-space as it seemed like it would result in relatively simple code for `check`, and be an easier transition from today's resource implementations.

But having a single request is worth considering, with the caveat that having everything in one request could mean less parallelism, since it all has to happen in one container with shared state.

Taking a look at today's `git` resource, which works by cloning the repo and running `git log`, it could actually work this way pretty easily by running `git log <branch>` for each space. This should even support `paths`/`ignore_paths`.

One big win would also be fewer containers running. Folks already complain about the number of `check` containers, and the current design could make that worse by having an additional container for each space.

So, I'll toy with this idea! Thanks!

Reply:
Great! I mentioned the `git-resource` but I what I was really thinking about was a `github-pr-resource` I've written myself, which uses the GraphQL api for Github and would reap 0 benefits for doing so if `check` was passed one space to check - so single request gets a +1 from me at least 😄